opsmngr/types.go

```diff
@@ -0,0 +1,41 @@
+// Copyright 2020 MongoDB Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package opsmngr
+
+import "encoding/json"
+
+// JSONArray is the type alias for []string
+type JSONArrayStr []string
```

Comment:
question around this change, if you do `curl` do you get an empty array from the api, `null`, or the field is not present?

Reply:
Unfortunately, it varies on the version of atlas/cm. For gov environment curl just doesn't return field if it's empty. For commercial it returns an empty array. It is crazy:(
So in the case of gov environment, we have null in ac structure. And we are trying to encode it as null... I've described it in the ticket.

Reply:
I believe it depends on atlas/cm version:
- atlas 18f1b47fec@master - returns empty array
- atlas 5404767157@v20201124 - removes the field
- cm 4.4.5.103.20201104T1729Z - returns empty array

Reply:
I thought we can mark fields as `omitempty` though. Because we will have changes in config anyway (comparing with original data).

Reply:
yes, the reason behind my question is if we should do `omitempty` instead.
I ran this test for `MongoDBUser` adding `omitempty` and making the array a pointer, I think it covers your use case, let me know what you think 

```go
type MongoDBUser struct {
	AuthenticationRestrictions []string       `json:"authenticationRestrictions"`
	CustomData                 interface{}    `json:"customData,omitempty"`
	Database                   string         `json:"db"`
	InitPassword               string         `json:"initPwd,omitempty"` // The cleartext password to be assigned to the user
	Mechanisms                 *[]string      `json:"mechanisms,omitempty"`
	Password                   string         `json:"pwd,omitempty"`
	Roles                      []*Role        `json:"roles"`
	ScramSha256Creds           *ScramShaCreds `json:"scramSha256Creds,omitempty"`
	ScramSha1Creds             *ScramShaCreds `json:"scramSha1Creds,omitempty"`
	Username                   string         `json:"user"`
}

func TestAutomation_MongoDBUserNoMechanism(t *testing.T) {
	client, mux, teardown := setup()
	defer teardown()

	mux.HandleFunc(fmt.Sprintf("/groups/%s/automationConfig", projectID), func(w http.ResponseWriter, r *http.Request) {
		testMethod(t, r, http.MethodGet)
		_, _ = fmt.Fprint(w, `{
  "auth" : {
    "authoritativeSet" : false,
    "autoAuthMechanism" : "MONGODB-CR",
    "disabled" : true,
	"usersWanted": [
		{
			"db": "admin"
		}
	]
  }}`)
	})

	config, _, err := client.Automation.GetConfig(ctx, projectID)
	if err != nil {
		t.Fatalf("Automation.GetConfig returned error: %v", err)
	}

	expected := &AutomationConfig{
		Auth: Auth{
			AuthoritativeSet:  false,
			AutoAuthMechanism: "MONGODB-CR",
			Disabled:          true,
			Users: []*MongoDBUser{
				{
					Database: "admin",
				},
			},
		},
	}
	if diff := deep.Equal(config, expected); diff != nil {
		t.Error(diff)
	}
}

func TestAutomation_MongoDBUserEmptyMechanism(t *testing.T) {
	client, mux, teardown := setup()
	defer teardown()

	mux.HandleFunc(fmt.Sprintf("/groups/%s/automationConfig", projectID), func(w http.ResponseWriter, r *http.Request) {
		testMethod(t, r, http.MethodGet)
		_, _ = fmt.Fprint(w, `{
  "auth" : {
    "authoritativeSet" : false,
    "autoAuthMechanism" : "MONGODB-CR",
    "disabled" : true,
	"usersWanted": [
		{
			"db": "admin",
			"mechanisms": []	
		}
	]
  }}`)
	})

	config, _, err := client.Automation.GetConfig(ctx, projectID)
	if err != nil {
		t.Fatalf("Automation.GetConfig returned error: %v", err)
	}

	expected := &AutomationConfig{
		Auth: Auth{
			AuthoritativeSet:  false,
			AutoAuthMechanism: "MONGODB-CR",
			Disabled:          true,
			Users: []*MongoDBUser{
				{
					Database:   "admin",
					Mechanisms: &[]string{},
				},
			},
		},
	}
	if diff := deep.Equal(config, expected); diff != nil {
		t.Error(diff)
	}
}

func TestAutomation_UpdateMongoDBUserEmptyMechanism(t *testing.T) {
	client, mux, teardown := setup()
	defer teardown()

	updateRequest := &AutomationConfig{
		Auth: Auth{
			AuthoritativeSet:  false,
			AutoAuthMechanism: "MONGODB-CR",
			Disabled:          true,
			Users: []*MongoDBUser{
				{
					Database:   "admin",
					Mechanisms: &[]string{},
				},
				{
					Database: "admin",
				},
			},
		},
	}
	mux.HandleFunc(fmt.Sprintf("/groups/%s/automationConfig", projectID), func(w http.ResponseWriter, r *http.Request) {
		expected := map[string]interface{}{
			"auth": map[string]interface{}{
				"authoritativeSet":     false,
				"autoAuthMechanism":    "MONGODB-CR",
				"autoAuthRestrictions": interface{}(nil),
				"disabled":             true,
				"usersDeleted":         interface{}(nil),
				"usersWanted": []interface{}{
					map[string]interface{}{
						"authenticationRestrictions": interface{}(nil),
						"db":                         "admin",
						"roles":                      interface{}(nil),
						"user":                       "",
						"mechanisms":                 []interface{}{},
					},
					map[string]interface{}{
						"authenticationRestrictions": interface{}(nil),
						"db":                         "admin",
						"roles":                      interface{}(nil),
						"user":                       "",
					},
				},
			},
			"backupVersions":       interface{}(nil),
			"balancer":             interface{}(nil),
			"cpsModules":           interface{}(nil),
			"indexConfigs":         interface{}(nil),
			"mongosqlds":           interface{}(nil),
			"mongots":              interface{}(nil),
			"onlineArchiveModules": interface{}(nil),
			"options":              interface{}(nil),
			"processes":            interface{}(nil),
			"replicaSets":          interface{}(nil),
			"roles":                interface{}(nil),
			"sharding":             interface{}(nil),
		}
		var v map[string]interface{}
		err := json.NewDecoder(r.Body).Decode(&v)
		if err != nil {
			t.Fatalf("decode json: %v", err)
		}
		t.Logf("v=%#v\n", v)
		if diff := deep.Equal(v, expected); diff != nil {
			t.Error(diff)
		}
		_, _ = fmt.Fprint(w, `{}`)
	})

	_, err := client.Automation.UpdateConfig(ctx, projectID, updateRequest)
	if err != nil {
		t.Fatalf("Automation.UpdateConfig returned error: %v", err)
	}
}
```
And they pass

Reply:
you can also see a bunch of keys that get serialized when null so we can expand these testes for your use case

Reply:
Yes, your kungfu is better then mine:) Thanks a lot for the advice! I've added your variant with tests for this case.